### ISSUE TITLE
pre-commit hook argument for schemafile doesn't support ~/

### ISSUE THREAD

**Issue by dvankampen** [sirosen/check-jsonschema#9]
If we have a schemafile in the user's home directory, I can pass `~/schemafile.json` when invoking `check-jsonschema` from the command line without error, but if I pass `args: ["--schemafile", "~/schemafile.json"]` to the pre-commit hook, I get a `FileNotFoundError`, because it does not expand the ~/ operator.  Looks like the call to `open` could use a `os.path.expanduser` on the path being opened?  This allows multiple developers with different usernames to have the schemafile in their home directory, without having to have the username hard-coded in the pre-commit yaml file.

**Comment by Borda** [comment]
you can try to replace `~` by `$HOME`

**Comment by dvankampen** [comment]
Just tried `$HOME`, tried `${HOME}`, tried `\$HOME` (in case the $ needed escaping), they all produce a `FileNotFoundError`.  It is because it is simply using an `open` call 
![image](https://user-images.githubusercontent.com/2799288/127857802-2fd4e8f1-f1e7-4d3b-987b-762de70cea5b.png)
which doesn't support environment variables or bash conventions like ~.

**Comment by Borda** [comment]
right, there shall be `os.path.expanduser(args.schemafile)`

**Comment by sirosen** [comment]
I think there's a fair case to be made that the current behavior is correct. The shell -- or whatever tool is calling `check-jsonschema` -- is responsible for any env var or tilde expansion prior to passing args.
I'm not _strongly_ opposed to adding an expanduser call, as it does no harm in 99.9% of cases, but before I add anything I want to check if a simple change in your usage will suffice.

`pre-commit` does not do any environment munging -- in fact, [the maintainer has been pretty clear about that](https://github.com/pre-commit/pre-commit/issues/758#issuecomment-404554291) -- so `$HOME` should be preserved. Rather than having `check-jsonschema` do the expansion for you, why not ask an expert, `bash`?
Simply replace the `entry` with a `bash` invocation which expands `$HOME` for you. e.g.
```yaml
- repo: https://github.com/sirosen/check-jsonschema
  rev: 0.4.0
  hooks:
    - id: check-jsonschema
      files: foo.json
      entry: bash -c 'check-jsonschema --schemafile "$HOME/.cache/jsonschema_validate/github-workflow" "$@"' --
```

I will warn you to be cautious in general relying on uncontrolled data in people's homedirs. It strikes me as likely to cause issues -- but your workflow, your rules. 😄

As an alternative workaround, there's always the "escape valve" of adding a local script hook and putting whatever logic you want into a bash script. Just pass `$@` through to `check-jsonschema` and you can still have `pre-commit` select which filenames to pass.

If neither of these work, we can consider `expanduser`, but I'd rather avoid making that change.

**Comment by dvankampen** [comment]
@sirosen  thank you so much for the well thought out reply!  Your suggestions seem promising.  The issue I have now with `"$@"` is that it doesn't handle filenames with whitespace, but I can track that down, I think.

**Comment by sirosen** [comment]
I haven't heard more about this, but I'm confident that this can be solved without a change in `check-jsonschema`, at least for `pre-commit` usage.

That said, I'm looking back at this and think I'm being too rigid. (I often find it frustrating when maintainers choose "purity" rather being pragmatic and helping me. Now _I'm_ being that guy. 😂 )

`os.path.expanduser` is documented as not doing anything weird, and I'd rather do that than `expandvars` (`$HOME` support) because it could _definitely_ produce some surprising results.
I'll make this change and get started on a testsuite for `check-jsonschema`, perhaps as early as this weekend.

**Comment by sirosen** [comment]
I've released v0.5.0 . It adds a call to `expanduser()` which should allow you to call
```
$ check-jsonschema --schemafile '~/foo.json' ...
```
and see the `~` expand.

0.5.0 was a fairly extensive refactor, aimed at making more of the code testable. There are also some new flags (`--no-cache` and `--default-filetype`), which will hopefully be useful. Please open issues if you see any problems!
